File: Mbs.Api.Host.Ng/ClientApp/src/app/shared/mbs/instruments/instrument.ts
import { InstrumentType } from './types/instrument-type.enum';
import { ExchangeMic } from './../markets/exchange-mic.enum';
import { Stock } from './types/stock';
import { Etf } from './types/etf';
import { Etv } from './types/etv';
import { Inav } from './types/inav';
import { Fund } from './types/fund';
import { Index } from './types';

/** Contains information to describe an instrument. */
export class Instrument {
  /** An optional symbol (ticker) of the security. */
  symbol?: string;

  /** A name of the instrument. */
  name?: string;

  /** An optional textual description for the instrument. */
  description?: string;

  /** The instrument type. */
  type!: InstrumentType;

  /** An exchange representations according to ISO 10383 Market Identifier Code (MIC). */
  mic!: ExchangeMic;

  /** An *ISIN*. */
  isin?: string;

  /** An additional information for stocks. */
  stock?: Stock;

  /** An additional information for Exchange-Traded Vehicles. */
  etv?: Etv;

  /** An additional information for Exchange-Traded Funds. */
  etf?: Etf;

  /** An additional information for *Indicative Net Asset Values*. */
  inav?: Inav;

  /** An additional information for funds. */
  fund?: Fund;

  /** An additional information for indices. */
  index?: Index;

  constructor(data?: Instrument) {
    if (data) {
      for (const property in data) {
        if (data.hasOwnProperty(property)) {
          (this as any)[property] = (data as any)[property];
        }
      }
      this.stock = data.stock && !(data.stock as any).toJSON ? new Stock(data.stock) : (this.stock as Stock);
      this.etv = data.etv && !(data.etv as any).toJSON ? new Etv(data.etv) : (this.etv as Etv);
      this.etf = data.etf && !(data.etf as any).toJSON ? new Etf(data.etf) : (this.etf as Etf);
      this.inav = data.inav && !(data.inav as any).toJSON ? new Inav(data.inav) : (this.inav as Inav);
      this.fund = data.fund && !(data.fund as any).toJSON ? new Fund(data.fund) : (this.fund as Fund);
      this.index = data.index && !(data.index as any).toJSON ? new Index(data.index) : (this.index as Index);
    }
    if (!data) {
      this.type = InstrumentType.Undefined;
      this.mic = ExchangeMic.Xxxx;
    }
  }

  toJSON(data?: any) {
    data = typeof data === 'object' ? data : {};
    // tslint:disable:no-string-literal
    data['Symbol'] = this.symbol;
    data['Name'] = this.name;
    data['Description'] = this.description;
    data['Type'] = this.type;
    data['Mic'] = this.mic;
    data['Isin'] = this.isin;
    data['Stock'] = this.stock ? this.stock.toJSON() : (undefined as any);
    data['Etv'] = this.etv ? this.etv.toJSON() : (undefined as any);
    data['Etf'] = this.etf ? this.etf.toJSON() : (undefined as any);
    data['Inav'] = this.inav ? this.inav.toJSON() : (undefined as any);
    data['Fund'] = this.fund ? this.fund.toJSON() : (undefined as any);
    data['Index'] = this.index ? this.index.toJSON() : (undefined as any);
    // tslint:enable:no-string-literal
    return data;
  }
}

File: Mbs.Api/Models/Trading/Instruments/Instrument.cs
using System.Collections.Generic;
using System.ComponentModel;
using System.ComponentModel.DataAnnotations;
using System.Runtime.Serialization;
using Mbs.Trading.Instruments;
using Mbs.Trading.Markets;

namespace Mbs.Api.Models.Trading.Instruments
{
    /// <summary>
    /// Contains information to describe an instrument.
    /// </summary>
    [DataContract]
    public class Instrument : IValidatableObject
    {
        /// <summary>
        /// Gets or sets the optional symbol (ticker) of the security.
        /// </summary>
        [DataMember(IsRequired = false, EmitDefaultValue = false)]
        public string Symbol { get; set; }

        /// <summary>
        /// Gets or sets a name of the instrument.
        /// </summary>
        [DataMember(IsRequired = false, EmitDefaultValue = false)]
        public string Name { get; set; }

        /// <summary>
        /// Gets or sets an optional textual description for the instrument.
        /// </summary>
        [DataMember(IsRequired = false, EmitDefaultValue = false)]
        public string Description { get; set; }

        /// <summary>
        /// Gets or sets the instrument type.
        /// </summary>
        [DataMember(IsRequired = false, EmitDefaultValue = true)]
        [DefaultValue(InstrumentType.Undefined)]
        public InstrumentType Type { get; set; } = InstrumentType.Undefined;

        /// <summary>
        /// Gets or sets an exchange representations according to ISO 10383 Market Identifier Code (MIC).
        /// </summary>
        [DataMember(IsRequired = false, EmitDefaultValue = true)]
        [DefaultValue(ExchangeMic.Xxxx)]
        public ExchangeMic Mic { get; set; } = ExchangeMic.Xxxx;

        /// <summary>
        /// Gets or sets an ISIN.
        /// </summary>
        [DataMember(IsRequired = false, EmitDefaultValue = false)]
        public string Isin { get; set; }

        /// <summary>
        /// Gets or sets an additional information for stocks.
        /// </summary>
        [DataMember(IsRequired = false, EmitDefaultValue = false)]
        public Stock Stock { get; set; }

        /// <summary>
        /// Gets or sets an additional information for Exchange-Traded Vehicles.
        /// </summary>
        [DataMember(IsRequired = false, EmitDefaultValue = false)]
        public Etv Etv { get; set; }

        /// <summary>
        /// Gets or sets an additional information for Exchange-Traded Funds.
        /// </summary>
        [DataMember(IsRequired = false, EmitDefaultValue = false)]
        public Etf Etf { get; set; }

        /// <summary>
        /// Gets or sets an additional information for Indicative Net Asset Values.
        /// </summary>
        [DataMember(IsRequired = false, EmitDefaultValue = false)]
        public Inav Inav { get; set; }

        /// <summary>
        /// Gets or sets an additional information for funds.
        /// </summary>
        [DataMember(IsRequired = false, EmitDefaultValue = false)]
        public Fund Fund { get; set; }

        /// <summary>
        /// Gets or sets an additional information for indices.
        /// </summary>
        [DataMember(IsRequired = false, EmitDefaultValue = false)]
        public Index Index { get; set; }

        /// <inheritdoc />
        public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
        {
            if (Stock != null && Type != InstrumentType.Stock)
            {
                yield return new ValidationResult($"The Stock field is not allowed when the type is {Type}.", new[] { "Type", "Stock" });
            }

            if (Etv != null && Type != InstrumentType.Etv)
            {
                yield return new ValidationResult($"The Etv field is not allowed when the type is {Type}.", new[] { "Type", "Etv" });
            }

            if (Etf != null && Type != InstrumentType.Etf)
            {
                yield return new ValidationResult($"The Etf field is not allowed when the type is {Type}.", new[] { "Type", "Etf" });
            }

            if (Inav != null && Type != InstrumentType.Inav)
            {
                yield return new ValidationResult($"The Inav field is not allowed when the type is {Type}.", new[] { "Type", "Inav" });
            }

            if (Fund != null && Type != InstrumentType.Fund)
            {
                yield return new ValidationResult($"The Fund field is not allowed when the type is {Type}.", new[] { "Type", "Fund" });
            }

            if (Index != null && Type != InstrumentType.Index)
            {
                yield return new ValidationResult($"The Index field is not allowed when the type is {Type}.", new[] { "Type", "Index" });
            }
        }
    }
}
